### PR TITLE
feat: Eng 374 convert newsletter server action to trpc

### PR DESCRIPTION
**Pull request by MichaelUnkey**
## What does this PR do?

Changed newsletter email signup from Server Actions to tRPC.

Fixes # (issue)

*If there is not an issue for this, please create one first. This is used to tracking purposes and also helps use understand why this PR exists*

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

Make sure newsletter signup still functions as expected with resend. 

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [Contributing Guide](./CONTRIBUTING.md)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand areas
- [x] Ran `pnpm build`
- [x] Ran `pnpm fmt`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues

### Appreciated

- [ ] If a UI change was made: Added a screen recording or screenshots to this PR
- [ ] Updated the Unkey Docs if changes were necessary

### CODE REVIEW

File: apps/web/components/landing/newsletter.tsx [hunk 1]
```diff
@@ -36,28 +35,34 @@ const SubmitButton = () => {
 export function NewsletterForm() {
   const formRef = useRef<HTMLFormElement>(null);
   const { toast } = useToast();
-  return (
-    <form
-      ref={formRef}
-      className="max-w-md"
-      action={async (formData) => {
-        const test = await addEmail(formData);
+  function handleSubmit(event: any) {
```
Comment (chronark) [inline apps/web/components/landing/newsletter.tsx:38]: Let's make this typesafe similar to here: https://github.com/unkeyed/unkey/blob/f6174806463e2e96d8120f3cf497de5588ec9987/apps/web/app/(authenticated)/(app)/app/settings/team/invite.tsx#L56

File: apps/web/lib/trpc/routers/newsletter.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+      if (!RESEND_API_KEY || !RESEND_AUDIENCE_ID) {
+        throw new TRPCError({
+          code: "INTERNAL_SERVER_ERROR",
+          message: "An unexpected error occurred, please try again later.",
+        });
+      }
+
+      const email = input.email;
+      if (!email) {
```
Comment (chronark) [inline apps/web/lib/trpc/routers/newsletter.ts:25]: we don't need this check, by using a zod schema in the input function above, it is already checked

File: apps/web/lib/trpc/routers/newsletter.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import { TRPCError } from "@trpc/server";
+import { Resend } from "resend";
+import { z } from "zod";
+import { t } from "../trpc";
+
+export const newsletterRouter = t.router({
+  signup: t.procedure
+    .input(
+      z.object({
+        email: z.string(),
```
Comment (chronark) [inline apps/web/lib/trpc/routers/newsletter.ts:11]: they got a nice utility that does some validity checking for us
```suggestion
        email: z.string().email(),
```

File: apps/web/lib/trpc/routers/newsletter.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+          message: "Email is not valid. Please try again.",
+        });
+      }
+      const resend = new Resend(RESEND_API_KEY);
+      const res = await resend.contacts.create({
+        audience_id: RESEND_AUDIENCE_ID,
+        email: email,
+      });
+      if (res.error) {
+        throw new TRPCError({
```
Comment (chronark) [inline apps/web/lib/trpc/routers/newsletter.ts:37]: let's log the error here
```suggestion
        console.error(res.error)
        throw new TRPCError({
```

**Review by chronark — CHANGES_REQUESTED** [review]

**Comment by MichaelUnkey** [comment]
Issues with middleware moving back to server actions

**Pull request closed without merging**